crates/autopilot/src/domain/settlement/encoded/mod.rs

```diff
@@ -0,0 +1,249 @@
+
+        let (calldata, metadata) = data.split_at(data.len() - Self::META_DATA_LEN);
+        let metadata: Option<[u8; Self::META_DATA_LEN]> = metadata.try_into().ok();
+        let auction_id = metadata
+            .map(auction::Id::from_be_bytes)
+            .ok_or(Error::MissingAuctionId)?;
+
+        let tokenized = function
+            .decode_input(calldata)
+            .map_err(|err| EncodingError::Decoding(err).with(auction_id))?;
+        let tokenized = <tokenized::Settlement>::from_token(web3::ethabi::Token::Tuple(tokenized))
+            .map_err(|err| EncodingError::Tokenizing(err).with(auction_id))?;
+
+        let (tokens, clearing_prices, decoded_trades, interactions) = tokenized;
+
+        let mut trades = Vec::with_capacity(decoded_trades.len());
+        for trade in decoded_trades {
+            let flags = TradeFlags(trade.8);
+            let signature = boundary::Signature::from_bytes(flags.signing_scheme(), &trade.10 .0)
+                .map_err(boundary::order::Error::Signature)
+                .map_err(|err| EncodingError::OrderUidRecover(err).with(auction_id))?;
+
+            trades.push(Trade {
+                order_uid: boundary::order::order_uid(&trade, &tokens, &domain_separator)
+                    .map_err(|err| EncodingError::OrderUidRecover(err).with(auction_id))?,
+                sell_token_index: trade.0.as_usize(),
+                buy_token_index: trade.1.as_usize(),
+                receiver: trade.2.into(),
+                sell_amount: trade.3.into(),
+                buy_amount: trade.4.into(),
+                valid_to: trade.5,
+                app_data: order::AppDataHash(trade.6 .0),
+                fee_amount: trade.7.into(),
+                flags,
+                executed: trade.9.into(),
+                signature: signature.into(),
+            })
+        }
+        let tokens: Vec<eth::Address> = tokens.into_iter().map(Into::into).collect();
+        let interactions = interactions.map(|inner| inner.into_iter().map(Into::into).collect());
+
+        Ok(Self {
+            tokens,
+            clearing_prices,
+            trades,
+            interactions,
+            auction_id,
+        })
+    }
+
+    pub fn auction_id(&self) -> auction::Id {
+        self.auction_id
+    }
+
+    pub fn trades(&self) -> &[Trade] {
+        &self.trades
+    }
+
+    pub fn tokens(&self) -> &[eth::Address] {
+        &self.tokens
+    }
+
+    pub fn clearing_prices(&self) -> &[eth::U256] {
+        &self.clearing_prices
+    }
+}
+
+#[derive(Debug)]
+pub struct Trade {
+    pub sell_token_index: usize,
+    pub buy_token_index: usize,
+    pub receiver: eth::Address,
+    pub sell_amount: eth::TokenAmount,
+    pub buy_amount: eth::TokenAmount,
+    pub valid_to: u32,
+    pub app_data: order::AppDataHash,
+    pub fee_amount: eth::TokenAmount,
+    pub flags: TradeFlags,
+    pub executed: eth::TargetAmount,
+    pub signature: order::Signature,
```

Comment:
Should these fields be public? They don't feel very domain-y (but instead very implementation specific to our settlement contract)

I think for instance `sell_token_index` and `buy_token_index` should be abstracted in a way that one can get the price directly from the trade in some way (ie with a reference to the originating settlement)

Reply:
private and cleaned up